src/catalog/catalog.cpp

```diff
@@ -30,413 +30,764 @@ Catalog *Catalog::GetInstance(void) {
-      bool has_primary_key = false;
-      auto &schema_columns = table->GetSchema()->GetColumns();
-      for (auto &column : schema_columns)
-        if (column.IsPrimary()) {
-          has_primary_key = true;
-          break;
-        }
-      if (has_primary_key == true)
-        CreatePrimaryIndex(database_name, table_name);
-
-      // Update catalog_table with this table info
-      auto tuple =
-          GetTableCatalogTuple(
-              databases_[START_OID]->GetTableWithName(TABLE_CATALOG_NAME)->GetSchema(),
-              table_id, table_name, database_id, database->GetDBName(), pool_);
-      // Another way of insertion using transaction manager
-      catalog::InsertTuple(
-          databases_[START_OID]->GetTableWithName(TABLE_CATALOG_NAME),
-          std::move(tuple), txn);
-      return ResultType::SUCCESS;
+    auto database = GetDatabaseWithOid(database_oid);
+
+    // Check duplicate column names
+    std::set<std::string> column_names;
+    auto columns = schema.get()->GetColumns();
+
+    for (auto column : columns) {
+      auto column_name = column.GetName();
+      if (column_names.count(column_name) == 1) {
+        LOG_TRACE("Found a column with the same name. Return RESULT_FAILURE");
+        return ResultType::FAILURE;
+      }
+      column_names.insert(column_name);
+    }
+
+    // Create actual table
+    auto pg_table = TableCatalog::GetInstance();
+    table_oid = pg_table->GetNextOid();
+    bool own_schema = true;
+    bool adapt_table = false;
+    auto table = storage::TableFactory::GetDataTable(
+        database_oid, table_oid, schema.release(), table_name,
+        DEFAULT_TUPLES_PER_TILEGROUP, own_schema, adapt_table);
+    database->AddTable(table);
+
+    // Update pg_table with this table info
+    pg_table->InsertTable(table_oid, table_name, database_oid, pool_.get(),
+                          txn);
+
+    for (auto column : table->GetSchema()->GetColumns()) {
+      ColumnCatalog::GetInstance()->InsertColumn(
+          table_oid, column.GetName(), column.GetOffset(), column.GetType(),
+          column.IsInlined(), column.GetConstraints(), pool_.get(), txn);
     }
+
+    CreatePrimaryIndex(database_name, table_name, txn);
+
+    if (single_statement_txn) {
+      txn_manager.CommitTransaction(txn);
+    }
+
+    return ResultType::SUCCESS;
   } catch (CatalogException &e) {
-    LOG_ERROR("Could not find a database with name %s", database_name.c_str());
+    LOG_TRACE("Can't found database. Return RESULT_FAILURE");
     return ResultType::FAILURE;
   }
 }
 
 ResultType Catalog::CreatePrimaryIndex(const std::string &database_name,
-    const std::string &table_name) {
+                                       const std::string &table_name,
+                                       concurrency::Transaction *txn) {
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  bool single_statement_txn = false;
+
+  if (txn == nullptr) {
```

Comment:
Same comment as above. Should we just force everyone to invoke these methods in a txn?